Add bech32 encoding and decoding funcs

george-larionov opened the pull request:
Resolves: #40381

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added Bech32 and Bech32m encoding and decoding functions (issue #40381).

### Documentation entry for user-facing changes
- [x] Documentation is written (mandatory for new features)

george-larionov committed:
adding test files

george-larionov committed:
adding basic tests for bech32 encode and decode

george-larionov committed:
removing some whitespace

george-larionov committed:
adding bech32 algorithm code reference impl

george-larionov committed:
adding decoder function, currently outputting 5-bit address instead of 8-bit

george-larionov committed:
added 5->8 bit conversion and working testcases

george-larionov committed:
cleaning up dupe code

george-larionov committed:
added encoding, currently outputting slightly incorrect values, something wrong with the checksum

george-larionov committed:
working implementation with tests

george-larionov committed:
removing unnecessary includes

george-larionov committed:
removing unneeded dependency file

george-larionov committed:
code cleanup and updating tests

george-larionov committed:
more code cleanup

george-larionov committed:
fixing issue with fixed strings and padding

george-larionov committed:
even more code cleanup

george-larionov committed:
fixing issue with fixed strings for encoding and updating tests

george-larionov committed:
adding doc

george-larionov committed:
updating doc

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=80239) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=80239) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=80239) it.</sub>

george-larionov committed:
addressing compiler warnings

george-larionov committed:
updating tests and dictionary

george-larionov committed:
changes to appease style check

george-larionov committed:
removing empty line to satisfy style checker

george-larionov committed:
moving pragma to before comment

george-larionov committed:
for some reason style checker doesn't like namespace comments

glarik commented:
There are a couple of failing tests that seem unrelated to my changes, can I merge with these failures or should I address them somehow?

vitlibar commented on docs/en/sql-reference/functions/encoding-functions.md:
````diff
@@ -967,3 +967,122 @@ Result:
+**Syntax**
+
+```sql
+Bech32Encode(hrp, data, witver(optional))
````
Usually we use brackets to tell that a parameter is optional, for example [see](https://clickhouse.com/docs/sql-reference/functions/date-time-functions#makedatetime64)

vitlibar replied:
Maybe it makes sense to make `hrp` an optional parameter too with `bc` as its default value, WDYT? Also it looks more correct to use `1` as the default value for `witver` (because `Bech32` has a vulnerability and `Bech32m` is more secure).

glarik replied:
I agree about the default for witver, I will change it to 1. For hrp, I don't think a default makes sense because we don't know whether the user wants a Mainnet or Testnet address, and additionally there are other applications of Bech32, such as for different coins, that use different HRPs (e.g. Litecoin: "ltc"). I think we should leave the Bech32 functions as generic implementations of the algorithm, which is not Bitcoin-specific.

vitlibar commented:
> I chose not to link the Bech32 repository, and instead simply added the files needed directly to the Clickhouse codebase

I think it's better to copy these files to the contrib directory (without making a submodule, just make a folder) because the `src/Functions` folder is quite overloaded. Please also add some readme file containing information about where you copied these files originally from and also put the license file to that folder. We have script `list-licenses.sh` which scans our repository to find all licenses to show them in table `system.licenses`.

vitlibar commented on tests/queries/0_stateless/03512_bech32_enc_dec.sql:
```diff
@@ -0,0 +1,150 @@
+('tb', '8f8cdd4364bb7dca11c49743da2c4b54062fa0388bbf924078', 1),
+('bc', '50b80d45cc275f36eb5fb2c22a93f6a4e83ba9380e55c67f6a', 15),
+('tb', 'b103a1937c6e2fb9de707a4be02d5d39e217b4bca7ce3c9c12', 0),
+('bcrt', '95eb334ff82ef8ad76151c29094abdae6c9e8bb8244523e347', 2);
```
Please check if the following query works:
```
SELECT bech32Encode('bc', unhex(data)) FROM hex_data;
```

glarik replied:
Wow, I wouldn't have caught that, thanks! I have updated the code to handle mixed const and column data.

glarik commented on src/Functions/FunctionsBech32Representation.cpp:
```diff
@@ -0,0 +1,470 @@
+
+using bech32_data = std::vector<uint8_t>;
+
+/** Convert from one power-of-2 number base to another. */
```
Sure, I'll add an example. Pad is whether to pad the output or not. When going from 8-bits -> 5-bits we are sometimes left with some "extra" bits, like 1 8-bit byte must be mapped to 2 5-bit bytes, so there's 2 extra bits left over, which we pad with zeros. This is important so that we don't lose info during encoding, and when decoding we simply throw those extra bits away. I will include this info in the comment.

george-larionov committed:
updating comments for convertbits

george-larionov committed:
changing default algo version and updating docs

george-larionov committed:
moving bech32 impl to contrib folder and updating tests after changing default algo in previous commit

rschu1ze committed:
Merge remote-tracking branch 'ClickHouse/master' into add_bech32_enc_dec

george-larionov committed:
changing dummy values to nullptrs

george-larionov committed:
handling mixed const and column data

george-larionov committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into add_bech32_enc_dec

george-larionov committed:
removing trailing whitespaces

george-larionov committed:
adding readme

glarik commented:
@vitlibar I believe I have addressed all of your comments, please take another look when you get the chance.

rschu1ze committed:
Minor cleanups

rschu1ze reviewed (approved):
Made some additional minor changes:
- functions `bechEncode|Decode` are now conditionally build, depending on CMake variable `ENABLE_LIBRARIES`
- the function is no longer case-insensitive. We make functions usually only case-insensitive for compatibility reasons with dinosaur databases like MySQL. In this case, the new functionality is too specialized to bother about compatibility.
- some whitespace changes for consistency with the rest of the codebase

george-larionov committed:
updating tests and dictionary

rschu1ze committed:
Fix tests + simplify naming of implementation file

rschu1ze committed:
Merge remote-tracking branch 'ClickHouse/master' into add_bech32_enc_dec

vitlibar commented on src/Functions/bech32.cpp:
```diff
@@ -0,0 +1,488 @@
+        ColumnPtr col1 = arguments[1].column->convertToFullColumnIfConst();
+        ColumnPtr col2;
+        if (have_witver)
+            col2 = arguments[2].column;
```
The case when `witver` is a constant can be not working:
```
SELECT bech32Encode('bc', unhex(data), 1) FROM hex_data;
```

glarik replied:
I'm not sure what you mean, I have already included a test for a constant witver on 03512_bech32_functions.sql:63.
`SELECT bech32Encode(hrp, unhex(data), 1) FROM hex_data limit 1;`
This is slightly different from your test, but I think it tests the same thing, should I add the specific test you wrote?

george-larionov committed:
addressing fuzzer failure and updating test outputs

george-larionov committed:
Merge branch 'add_bech32_enc_dec' of github.com:glarik/ClickHouse into add_bech32_enc_dec

george-larionov committed:
Revert "addressing fuzzer failure and updating test outputs"

This reverts commit a1cd64800f9f9c04b8d700d6344fc763171bbece.

george-larionov committed:
adding check to ensure empty memory is never copied

george-larionov committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into add_bech32_enc_dec

alexey-milovidov commented:
@glarik, I'm testing it with addresses from this page: https://en.bitcoin.it/wiki/Bech32

It works for `bc1qc7slrfxkknqcq2jevvvkdgvrt8080852dfjewde450xdlk4ugp7szw5tk9` but does not work for `bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq`.

george-larionov commented:
@alexey-milovidov sorry for the late reply, I only noticed this comment recently, but you are right! My code had a bug, and the online bech32 calculator I was using to verify my results seems to also have the same issue! I have fixed it, and added that failing address as one of the tests. See the new PR here https://github.com/ClickHouse/ClickHouse/pull/84257.
